Refactor add/double operations

aztech-dev opened the pull request:
Original implementation checks that `(cmp(addend->x, augend->x) % curve prime) == 0`.

Considering that cmp returns `0` on equality, a (non specified) negative value if `a < b`, and a (non specified) positive value if `a > b`, `(cmp(addend->x, augend->x) % curve prime) == 0` is true in only three cases:

- `cmp(addend->x, augend->x) == 0`
- `cmp(addend->x, augend->x) == 0 - curve prime` (or any multiple of curve prime)
- `cmp(addend->x, augend->x) == curve prime` (or any multiple of curve prime)

Hence, it's safer to only check that `cmp(addend->x, augend->x) == 0`. I haven't checked the implementation of `gmp_cmp`, but it might as well return randomly a value equal to curve prime (or one of its multiples), leading the point addition code to believe `addend->x == augend->x` are equal, when in fact, they're not.

As an added bonus, it removes one math OP, making the code a very tiny little (really tiny) bit faster.

aztech-dev commented:
@afk11 Can you confirm my reasoning is correct ?

afk11 commented:
It appears you are right and this operation is not required. That's been in there since the first commit, but  seems to be the only instance of this bug 
`find . -type f -exec grep -F 'mod($math->cmp' {} +`

Since the code is explicitly testing that the `result % p == 0` the only chance there could be a problem is result was +/- i*p.. I'd guess that would make the odds of this affecting someone pretty slim :)

aztech-dev commented:
Nonetheless, slim is not null :p It actually depends on what cmp returns, but since the doc does not specify which negative/positive values are returned by gmp_cmp, let's not take the chance.

Actually, there's another commit coming, I refactored the add/double operations to make them easier to proof read, and extracted a modular math helper class to improve readability.

afk11 commented:
But since it can only be a problem when $math->cmp(x, x) == multiple*p - the probability decays with the size of your prime field ;) Glad you found it before someone ran into it!

btcdrak commented:
Was that an ACK from you @afk11?

aztech-dev commented:
Not sure he reviewed the latest commits  ...
Le 15 févr. 2015 13:32, "฿tcDrak" <notifications@github.com> a écrit :

> Was that an ACK from you @afk11 <https://github.com/afk11>?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mdanter/phpecc/pull/65#issuecomment-74415615>.
>

btcdrak commented:
Just for reference guys: I really prefer to see `ACK` when you agree something can be merged in your opinion.

afk11 commented:
It was an ACK from me. 

Interesting changes then also! Gonna raise an issue around InverseMod though, curious about something else we could do.

aztech-dev commented:
> Interesting changes then also! Gonna raise an issue around InverseMod though, curious about something else we could do.

Anything critical about this P/R or unrelated ? (asking as to whether @btcdrak can merge)

afk11 commented:
No all good for this PR, just was reminded of something that some of the EC curve operations can be optimized using coordinates on a

afk11 commented:
No all good for this PR, just was reminded of something that some of the EC curve operations can be optimized by mapping points onto a jacobian coordinate system. Not something I'd rush into implementing given the recent Point overhaul, just curious if it's something useful to look at?

aztech-dev commented:
Not sure about anything there, but it looks like the [Relic library](https://github.com/relic-toolkit/relic/blob/master/src/ep/relic_ep_add.c#L126) (which I used for reference along with Wikipedia and some other resources on EC math to refactor point addition) does implement it, so there may be some benefits.

I'll try to spill a rough implementation over the week, unless you want to give it a shot.
